# <center>Первичный анализ данных с Pandas</center>

In [ ]:
# Установка библиотек
# !pip install pandas
# !pip install numpy

In [1]:
import pandas as pd
import numpy as np

<img src="https://github.com/MalikaL17/course_materials/blob/main/img/creating_dataframe.png?raw=true">

In [2]:
from google.colab import drive
drive.mount('/content/drive')
#Путь к данным    'drive/My Drive/...'

Mounted at /content/drive


In [3]:
# чтение данных их csv-файла
df = pd.read_csv('drive/My Drive/Курс по анализу данных/Pandas/telecom_churn.csv') # read_excel()

In [ ]:
type(df)

pandas.core.frame.DataFrame

In [ ]:
df

State  Account length  ...  Customer service calls  Churn
0       KS             128  ...                       1  False
1       OH             107  ...                       1  False
2       NJ             137  ...                       0  False
3       OH              84  ...                       2  False
4       OK              75  ...                       3  False
...    ...             ...  ...                     ...    ...
3328    AZ             192  ...                       2  False
3329    WV              68  ...                       3  False
3330    RI              28  ...                       2  False
3331    CT             184  ...                       2  False
3332    TN              74  ...                       0  False

[3333 rows x 20 columns]

In [ ]:
type(df['Churn'])

pandas.core.series.Series

In [ ]:
# Создание датайрейма из словаря
tmp = pd.DataFrame({'id':[1,2,3,4], 'age':[32, 15, 76, 18]})
tmp 

id  age
0   1   32
1   2   15
2   3   76
3   4   18

**Посмотрим на размер данных, названия признаков и их типы**

In [ ]:
# вывод первых 5 заисей
df.head(3)
# вывод последних 5 заисей
df.tail(10)

State  Account length  ...  Customer service calls  Churn
3323    IN             117  ...                       5   True
3324    WV             159  ...                       1  False
3325    OH              78  ...                       2  False
3326    OH              96  ...                       1  False
3327    SC              79  ...                       2  False
3328    AZ             192  ...                       2  False
3329    WV              68  ...                       3  False
3330    RI              28  ...                       2  False
3331    CT             184  ...                       2  False
3332    TN              74  ...                       0  False

[10 rows x 20 columns]

In [ ]:
# размерность
df.shape

(3333, 20)

In [ ]:
# количество элементов
df.size

66660

In [ ]:
# Получить массив значений датафрейма
df.values

array([['KS', 128, 415, ..., 2.7, 1, False],
       ['OH', 107, 415, ..., 3.7, 1, False],
       ['NJ', 137, 415, ..., 3.29, 0, False],
       ...,
       ['RI', 28, 510, ..., 3.81, 2, False],
       ['CT', 184, 510, ..., 1.35, 2, False],
       ['TN', 74, 415, ..., 3.7, 0, False]], dtype=object)

In [4]:
# Посмотрим список колонок 
df.columns

Index(['State', 'Account length', 'Area code', 'International plan',
       'Voice mail plan', 'Number vmail messages', 'Total day minutes',
       'Total day calls', 'Total day charge', 'Total eve minutes',
       'Total eve calls', 'Total eve charge', 'Total night minutes',
       'Total night calls', 'Total night charge', 'Total intl minutes',
       'Total intl calls', 'Total intl charge', 'Customer service calls',
       'Churn'],
      dtype='object')

In [20]:
# Получить массив индексов
df.index

RangeIndex(start=0, stop=3333, step=1)

In [ ]:
# типы объектов встолбцах
df.dtypes

State                      object
Account length              int64
Area code                   int64
International plan         object
Voice mail plan            object
Number vmail messages     float64
Total day minutes         float64
Total day calls             int64
Total day charge          float64
Total eve minutes         float64
Total eve calls             int64
Total eve charge          float64
Total night minutes       float64
Total night calls           int64
Total night charge        float64
Total intl minutes        float64
Total intl calls            int64
Total intl charge         float64
Customer service calls      int64
Churn                        bool
dtype: object

In [ ]:
# общая информация о датафрейме
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3333 entries, 0 to 3332
Data columns (total 20 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   State                   3333 non-null   object 
 1   Account length          3333 non-null   int64  
 2   Area code               3333 non-null   int64  
 3   International plan      3333 non-null   object 
 4   Voice mail plan         3333 non-null   object 
 5   Number vmail messages   3331 non-null   float64
 6   Total day minutes       3333 non-null   float64
 7   Total day calls         3333 non-null   int64  
 8   Total day charge        3333 non-null   float64
 9   Total eve minutes       3333 non-null   float64
 10  Total eve calls         3333 non-null   int64  
 11  Total eve charge        3331 non-null   float64
 12  Total night minutes     3333 non-null   float64
 13  Total night calls       3333 non-null   int64  
 14  Total night charge      3333 non-null   

`bool`, `int64`, `float64` и `object` — это типы признаков. Видим, что 1 признак — логический (`bool`), 3 признака имеют тип `object` и 16 признаков — числовые.

**Изменить тип колонки** можно с помощью метода `astype`. Применим этот метод к признаку `Churn` и переведём его в `int64`:

In [ ]:
df["Churn"]

0       False
1       False
2       False
3       False
4       False
        ...  
3328    False
3329    False
3330    False
3331    False
3332    False
Name: Churn, Length: 3333, dtype: bool

In [ ]:
df["Churn"] = df["Churn"].astype("int64")

In [ ]:
df['Churn']

0       False
1       False
2       False
3       False
4       False
        ...  
3328    False
3329    False
3330    False
3331    False
3332    False
Name: Churn, Length: 3333, dtype: bool

Метод describe показывает основные статистические характеристики данных по каждому числовому признаку (типы int64 и float64): число непропущенных значений, среднее, стандартное отклонение, диапазон, медиану, 0.25 и 0.75 квартили.

In [ ]:
df

State  Account length  Area code International plan Voice mail plan  \
0       KS             128        415                 No             Yes   
1       OH             107        415                 No             Yes   
2       NJ             137        415                 No              No   
3       OH              84        408                Yes              No   
4       OK              75        415                Yes              No   
...    ...             ...        ...                ...             ...   
3328    AZ             192        415                 No             Yes   
3329    WV              68        415                 No              No   
3330    RI              28        510                 No              No   
3331    CT             184        510                Yes              No   
3332    TN              74        415                 No             Yes   

      Number vmail messages  Total day minutes  Total day calls  \
0                      25.0              265.1              110   
1                      26.0              161.6              123   
2                       0.0              243.4              114   
3                       0.0              299.4               71   
4                       0.0              166.7              113   
...                     ...                ...              ...   
3328                   36.0              156.2               77   
3329                    0.0              231.1               57   
3330                    0.0              180.8              109   
3331                    0.0              213.8              105   
3332                   25.0              234.4              113   

      Total day charge  Total eve minutes  Total eve calls  Total eve charge  \
0                45.07              197.4               99             16.78   
1                27.47              195.5              103             16.62   
2                41.38              121.2              110             10.30   
3                50.90               61.9               88              5.26   
4                28.34              148.3              122             12.61   
...                ...                ...              ...               ...   
3328             26.55              215.5              126             18.32   
3329             39.29              153.4               55             13.04   
3330             30.74              288.8               58             24.55   
3331             36.35              159.6               84             13.57   
3332             39.85              265.9               82             22.60   

      Total night minutes  Total night calls  Total night charge  \
0                   244.7                 91               11.01   
1                   254.4                103               11.45   
2                   162.6                104                7.32   
3                   196.9                 89                8.86   
4                   186.9                121                8.41   
...                   ...                ...                 ...   
3328                279.1                 83               12.56   
3329                191.3                123                8.61   
3330                191.9                 91                8.64   
3331                139.2                137                6.26   
3332                241.4                 77               10.86   

      Total intl minutes  Total intl calls  Total intl charge  \
0                   10.0                 3               2.70   
1                   13.7                 3               3.70   
2                   12.2                 5               3.29   
3                    6.6                 7               1.78   
4                   10.1                 3               2.73   
...                  ...               ...                ...   
3328                 9.9                 6            

In [ ]:
df.describe()

Account length    Area code  Number vmail messages  Total day minutes  \
count     3333.000000  3333.000000            3331.000000        3333.000000   
mean       101.064806   437.182418               8.103873         179.775098   
std         39.822106    42.371290              13.691036          54.467389   
min          1.000000   408.000000               0.000000           0.000000   
25%         74.000000   408.000000               0.000000         143.700000   
50%        101.000000   415.000000               0.000000         179.400000   
75%        127.000000   510.000000              20.000000         216.400000   
max        243.000000   510.000000              51.000000         350.800000   

       Total day calls  Total day charge  Total eve minutes  Total eve calls  \
count      3333.000000       3333.000000        3333.000000      3333.000000   
mean        100.435644         30.562307         200.980348       100.114311   
std          20.069084          9.259435          50.713844        19.922625   
min           0.000000          0.000000           0.000000         0.000000   
25%          87.000000         24.430000         166.600000        87.000000   
50%         101.000000         30.500000         201.400000       100.000000   
75%         114.000000         36.790000         235.300000       114.000000   
max         165.000000         59.640000         363.700000       170.000000   

       Total eve charge  Total night minutes  Total night calls  \
count       3331.000000          3333.000000        3333.000000   
mean          17.084014           200.872037         100.107711   
std            4.311873            50.573847          19.568609   
min            0.000000            23.200000          33.000000   
25%           14.160000           167.000000          87.000000   
50%           17.120000           201.200000         100.000000   
75%           20.000000           235.300000         113.000000   
max           30.910000           395.000000         175.000000   

       Total night charge  Total intl minutes  Total intl calls  \
count         3333.000000         3333.000000       3333.000000   
mean             9.039325           10.237294          4.479448   
std              2.275873            2.791840          2.461214   
min              1.040000            0.000000          0.000000   
25%              7.520000            8.500000          3.000000   
50%              9.050000           10.300000          4.000000   
75%             10.590000           12.100000          6.000000   
max             17.770000           20.000000         20.000000   

       Total intl charge  Customer service calls        Churn  
count        3333.000000             3333.000000  3333.000000  
mean            2.764581                1.562856     0.144914  
std             0.753773                1.315491     0.352067  
min             0.000000                0.000000     0.000000  
25%             2.300000                1.000000     0.000000  
50%             2.780000                1.000000     0.000000  
75%             3.270000                2.000000     0.000000  
max             5.400000                9.000000     1.000000

Чтобы посмотреть статистику по нечисловым признакам, нужно явно указать интересующие нас типы в параметре include. Можно также задать include='all', чтоб вывести статистику по всем имеющимся признакам.

In [ ]:
df.describe(include=["object", "bool"])

State International plan Voice mail plan
count   3333               3333            3333
unique    51                  2               2
top       WV                 No              No
freq     106               3010            2411

In [ ]:
df.describe(exclude=["int"])

State  Account length    Area code International plan Voice mail plan  \
count   3333     3333.000000  3333.000000               3333            3333   
unique    51             NaN          NaN                  2               2   
top       WV             NaN          NaN                 No              No   
freq     106             NaN          NaN               3010            2411   
mean     NaN      101.064806   437.182418                NaN             NaN   
std      NaN       39.822106    42.371290                NaN             NaN   
min      NaN        1.000000   408.000000                NaN             NaN   
25%      NaN       74.000000   408.000000                NaN             NaN   
50%      NaN      101.000000   415.000000                NaN             NaN   
75%      NaN      127.000000   510.000000                NaN             NaN   
max      NaN      243.000000   510.000000                NaN             NaN   

        Number vmail messages  Total day minutes  Total day calls  \
count             3331.000000        3333.000000      3333.000000   
unique                    NaN                NaN              NaN   
top                       NaN                NaN              NaN   
freq                      NaN                NaN              NaN   
mean                 8.103873         179.775098       100.435644   
std                 13.691036          54.467389        20.069084   
min                  0.000000           0.000000         0.000000   
25%                  0.000000         143.700000        87.000000   
50%                  0.000000         179.400000       101.000000   
75%                 20.000000         216.400000       114.000000   
max                 51.000000         350.800000       165.000000   

        Total day charge  Total eve minutes  Total eve calls  \
count        3333.000000        3333.000000      3333.000000   
unique               NaN                NaN              NaN   
top                  NaN                NaN              NaN   
freq                 NaN                NaN              NaN   
mean           30.562307         200.980348       100.114311   
std             9.259435          50.713844        19.922625   
min             0.000000           0.000000         0.000000   
25%            24.430000         166.600000        87.000000   
50%            30.500000         201.400000       100.000000   
75%            36.790000         235.300000       114.000000   
max            59.640000         363.700000       170.000000   

        Total eve charge  Total night minutes  Total night calls  \
count        3331.000000          3333.000000        3333.000000   
unique               NaN                  NaN                NaN   
top                  NaN                  NaN                NaN   
freq                 NaN                  NaN                NaN   
mean           17.084014           200.872037         100.107711   
std             4.311873            50.573847          19.568609   
min             0.000000            23.200000          33.000000   
25%            14.160000           167.000000          87.000000   
50%            17.120000           201.200000         100.000000   
75%            20.000000           235.300000         113.000000   
max            30.910000           395.000000         175.000000   

        Total night charge  Total intl minutes  Total intl calls  \
count          3333.000000         3333.000000       3333.000000   
unique                 NaN                 NaN               NaN   
top                    NaN                 NaN               NaN   
freq                   NaN                 NaN               NaN   
mean              9.039325           10.237294          4.479448   
std               2.275873            2.791840          2.461214   
min               1.040000            0.000000          0.000000   
25%               7.520000            8.500000          3.000000   
50%               9

### Сортировка

`DataFrame` можно отсортировать по значению какого-нибудь из признаков. В нашем случае, например, по `Total day charge` (`ascending=False` для сортировки по убыванию):

In [ ]:
df.sort_values(by="Total day charge", ascending=False).head()

State  Account length  Area code International plan Voice mail plan  \
365     CO             154        415                 No              No   
985     NY              64        415                Yes              No   
2594    OH             115        510                Yes              No   
156     OH              83        415                 No              No   
605     MO             112        415                 No              No   

      Number vmail messages  Total day minutes  Total day calls  \
365                     0.0              350.8               75   
985                     0.0              346.8               55   
2594                    0.0              345.3               81   
156                     0.0              337.4              120   
605                     0.0              335.5               77   

      Total day charge  Total eve minutes  Total eve calls  Total eve charge  \
365              59.64              216.5               94             18.40   
985              58.96              249.5               79             21.21   
2594             58.70              203.4              106             17.29   
156              57.36              227.4              116             19.33   
605              57.04              212.5              109             18.06   

      Total night minutes  Total night calls  Total night charge  \
365                 253.9                100               11.43   
985                 275.4                102               12.39   
2594                217.5                107                9.79   
156                 153.9                114                6.93   
605                 265.0                132               11.93   

      Total intl minutes  Total intl calls  Total intl charge  \
365                 10.1                 9               2.73   
985                 13.3                 9               3.59   
2594                11.8                 8               3.19   
156                 15.8                 7               4.27   
605                 12.7                 8               3.43   

      Customer service calls  Churn  
365                        1      1  
985                        1      1  
2594                       1      1  
156                        0      1  
605                        2      1

In [ ]:
# Сортировать можно и по группе столбцов:
df.sort_values(by=["Churn", "Total day charge"], ascending=[True, False]).head()

State  Account length  Area code International plan Voice mail plan  \
688     MN              13        510                 No             Yes   
2259    NC             210        415                 No             Yes   
534     LA              67        510                 No              No   
575     SD             114        415                 No             Yes   
2858    AL             141        510                 No             Yes   

      Number vmail messages  Total day minutes  Total day calls  \
688                    21.0              315.6              105   
2259                   31.0              313.8               87   
534                     0.0              310.4               97   
575                    36.0              309.9               90   
2858                   28.0              308.0              123   

      Total day charge  Total eve minutes  Total eve calls  Total eve charge  \
688              53.65              208.9               71             17.76   
2259             53.35              147.7              103             12.55   
534              52.77               66.5              123              5.65   
575              52.68              200.3               89             17.03   
2858             52.36              247.8              128             21.06   

      Total night minutes  Total night calls  Total night charge  \
688                 260.1                123               11.70   
2259                192.7                 97                8.67   
534                 246.5                 99               11.09   
575                 183.5                105                8.26   
2858                152.9                103                6.88   

      Total intl minutes  Total intl calls  Total intl charge  \
688                 12.1                 3               3.27   
2259                10.1                 7               2.73   
534                  9.2                10               2.48   
575                 14.2                 2               3.83   
2858                 7.4                 3               2.00   

      Customer service calls  Churn  
688                        3      0  
2259                       3      0  
534                        4      0  
575                        1      0  
2858                       1      0

### Индексация и извлечение данных

#### Извлечение столбцов

Для извлечения отдельного столбца можно использовать конструкцию вида `DataFrame['Name']`. 

In [ ]:
df['Area code']

0       415
1       415
2       415
3       408
4       415
       ... 
3328    415
3329    415
3330    510
3331    510
3332    415
Name: Area code, Length: 3333, dtype: int64

Воспользуемся этим для ответа на вопрос: **какова доля нелояльных пользователей в нашем датафрейме?**

In [ ]:
df["Churn"].mean()

0.14491449144914492

14,5% — довольно плохой показатель для компании, с таким процентом оттока можно и разориться.

In [ ]:
# Достать несколько столбцов
df[['State','Area code', 'Total intl minutes']]

State  Area code  Total intl minutes
0       KS        415                10.0
1       OH        415                13.7
2       NJ        415                12.2
3       OH        408                 6.6
4       OK        415                10.1
...    ...        ...                 ...
3328    AZ        415                 9.9
3329    WV        415                 9.6
3330    RI        510                14.1
3331    CT        510                 5.0
3332    TN        415                13.7

[3333 rows x 3 columns]

In [ ]:
# слайсинг по строкам
df[0:10]

State  Account length  Area code International plan Voice mail plan  \
0    KS             128        415                 No             Yes   
1    OH             107        415                 No             Yes   
2    NJ             137        415                 No              No   
3    OH              84        408                Yes              No   
4    OK              75        415                Yes              No   
5    AL             118        510                Yes              No   
6    MA             121        510                 No             Yes   
7    MO             147        415                Yes              No   
8    LA             117        408                 No              No   
9    WV             141        415                Yes             Yes   

   Number vmail messages  Total day minutes  Total day calls  \
0                   25.0              265.1              110   
1                   26.0              161.6              123   
2                    0.0              243.4              114   
3                    0.0              299.4               71   
4                    0.0              166.7              113   
5                    0.0              223.4               98   
6                   24.0              218.2               88   
7                    0.0              157.0               79   
8                    0.0              184.5               97   
9                   37.0              258.6               84   

   Total day charge  Total eve minutes  Total eve calls  Total eve charge  \
0             45.07              197.4               99             16.78   
1             27.47              195.5              103             16.62   
2             41.38              121.2              110             10.30   
3             50.90               61.9               88              5.26   
4             28.34              148.3              122             12.61   
5             37.98              220.6              101             18.75   
6             37.09              348.5              108             29.62   
7             26.69              103.1               94              8.76   
8             31.37              351.6               80             29.89   
9             43.96              222.0              111             18.87   

   Total night minutes  Total night calls  Total night charge  \
0                244.7                 91               11.01   
1                254.4                103               11.45   
2                162.6                104                7.32   
3                196.9                 89                8.86   
4                186.9                121                8.41   
5                203.9                118                9.18   
6                212.6                118                9.57   
7                211.8                 96                9.53   
8                215.8                 90                9.71   
9                326.4                 97               14.69   

   Total intl minutes  Total intl calls  Total intl charge  \
0                10.0                 3               2.70   
1                13.7                 3               3.70   
2                12.2                 5               3.29   
3                 6.6                 7               1.78   
4                10.1                 3               2.73   
5                 6.3                 6               1.70   
6                 7.5                 7               2.03   
7                 7.1                 6               1.92   
8                 8.7                 4               2.35   
9                11.2                 5               3.02   

   Customer service calls  Churn  
0                       1      0  
1                       1      0  
2                       0      0  
3                       2      0  
4                       3      0  
5                       0      0  
6

Вытащить одну строку по индексу вот так `df[]` не получится. Для этого надо использовать специальный метод `loc`

In [ ]:
# Ошибка
#df[0]
# Использование метода loc
df.loc[0]

State                        KS
Account length              128
Area code                   415
International plan           No
Voice mail plan             Yes
Number vmail messages        25
Total day minutes         265.1
Total day calls             110
Total day charge          45.07
Total eve minutes         197.4
Total eve calls              99
Total eve charge          16.78
Total night minutes       244.7
Total night calls            91
Total night charge        11.01
Total intl minutes           10
Total intl calls              3
Total intl charge           2.7
Customer service calls        1
Churn                     False
Name: 0, dtype: object

Датафреймы можно индексировать как по названию столбца или строки, так и по порядковому номеру. Для индексации **по названию** используется метод **`loc`**, **по номеру** — **`iloc`**.

В первом случае мы говорим _«передай нам значения для id строк от 0 до 5 и для столбцов от State до Area code»_, а во втором — _«передай нам значения первых пяти строк в первых трёх столбцах»_. 

В случае `iloc` срез работает как обычно, однако в случае `loc` учитываются и начало, и конец среза. 

In [ ]:
tmp = pd.DataFrame({'a':[1,2,3],'b':[3,4,5]}, index=['ind_1','ind_2','ind_3'])
tmp

a  b
ind_1  1  3
ind_2  2  4
ind_3  3  5

In [ ]:
tmp.iloc[1]

a    2
b    4
Name: ind_2, dtype: int64

In [ ]:
tmp.loc['ind_1']

a    1
b    3
Name: ind_1, dtype: int64

In [ ]:
df.loc[0:5, "State":"Area code"]

State  Account length  Area code
0    KS             128        415
1    OH             107        415
2    NJ             137        415
3    OH              84        408
4    OK              75        415
5    AL             118        510

In [ ]:
df.loc[:,  "State":"Area code"]

State  Account length  Area code
0       KS             128        415
1       OH             107        415
2       NJ             137        415
3       OH              84        408
4       OK              75        415
...    ...             ...        ...
3328    AZ             192        415
3329    WV              68        415
3330    RI              28        510
3331    CT             184        510
3332    TN              74        415

[3333 rows x 3 columns]

In [ ]:
df.iloc[0:5, 0:3]

State  Account length  Area code
0    KS             128        415
1    OH             107        415
2    NJ             137        415
3    OH              84        408
4    OK              75        415

Часто бывает полезным достать **уникальные значения** из какого-то столбца

In [ ]:
df['State'].unique()

array(['KS', 'OH', 'NJ', 'OK', 'AL', 'MA', 'MO', 'LA', 'WV', 'IN', 'RI',
       'IA', 'MT', 'NY', 'ID', 'VT', 'VA', 'TX', 'FL', 'CO', 'AZ', 'SC',
       'NE', 'WY', 'HI', 'IL', 'NH', 'GA', 'AK', 'MD', 'AR', 'WI', 'OR',
       'MI', 'DE', 'UT', 'CA', 'MN', 'SD', 'NC', 'WA', 'NM', 'NV', 'DC',
       'KY', 'ME', 'MS', 'TN', 'PA', 'CT', 'ND'], dtype=object)

### Фильтрация данных

In [ ]:
# Выбрать всех клиентов, которые уйдут от компании
df[df['Churn']==1]

State  Account length  Area code International plan Voice mail plan  \
10      IN              65        415                 No              No   
15      NY             161        415                 No              No   
21      CO              77        408                 No              No   
33      AZ              12        408                 No              No   
41      MD             135        408                Yes             Yes   
...    ...             ...        ...                ...             ...   
3301    CA              84        415                 No              No   
3304    IL              71        510                Yes              No   
3320    GA             122        510                Yes              No   
3322    MD              62        408                 No              No   
3323    IN             117        415                 No              No   

      Number vmail messages  Total day minutes  Total day calls  \
10                      0.0              129.1              137   
15                      0.0              332.9               67   
21                      0.0               62.4               89   
33                      0.0              249.6              118   
41                     41.0              173.1               85   
...                     ...                ...              ...   
3301                    0.0              280.0              113   
3304                    0.0              186.1              114   
3320                    0.0              140.0              101   
3322                    0.0              321.1              105   
3323                    0.0              118.4              126   

      Total day charge  Total eve minutes  Total eve calls  Total eve charge  \
10               21.95              228.5               83             19.42   
15               56.59              317.8               97             27.01   
21               10.61              169.9              121             14.44   
33               42.43              252.4              119             21.45   
41               29.43              203.9              107             17.33   
...                ...                ...              ...               ...   
3301             47.60              202.2               90             17.19   
3304             31.64              198.6              140             16.88   
3320             23.80              196.4               77             16.69   
3322             54.59              265.5              122             22.57   
3323             20.13              249.3               97             21.19   

      Total night minutes  Total night calls  Total night charge  \
10                  208.8                111                9.40   
15                  160.6                128                7.23   
21                  209.6                 64                9.43   
33                  280.2                 90               12.61   
41                  122.2                 78                5.50   
...                   ...                ...                 ...   
3301                156.8                103                7.06   
3304                206.5                 80                9.29   
3320                120.1                133                5.40   
3322                180.5                 72                8.12   
3323                227.0                 56               10.22   

      Total intl minutes  Total intl calls  Total intl charge  \
10                  12.7                 6               3.43   
15                   5.4                 9               1.46   
21                   5.7                 6               1.54   
33                  11.8                 3               3.19   
41                  14.6                15               3.94   
...                  ...               ...                ...   
3301                10.4                 4            

Найдем ответа на вопрос: **каковы средние значения числовых признаков среди нелояльных пользователей?**

In [ ]:
df[df["Churn"] == 1].mean()

Account length            102.664596
Area code                 437.817805
Number vmail messages       5.115942
Total day minutes         206.914079
Total day calls           101.335404
Total day charge           35.175921
Total eve minutes         212.410145
Total eve calls           100.561077
Total eve charge           18.054969
Total night minutes       205.231677
Total night calls         100.399586
Total night charge          9.235528
Total intl minutes         10.700000
Total intl calls            4.163561
Total intl charge           2.889545
Customer service calls      2.229814
Churn                       1.000000
dtype: float64

Скомбинировав фильтрацию и индексацию, ответим на вопрос: **сколько в среднем в течение дня разговаривают по телефону нелояльные пользователи**?

In [ ]:
df[df["Churn"] == 1]["Total day minutes"].mean()

206.91407867494814

Несколько условий комбинируются через круглые скобочки логический оператор И / ИЛИ

In [ ]:
# & - логическое И
df[(df["Total day minutes"] >= 20) & (df["Total day minutes"] < 200)] 
# | - логическое ИЛИ
df[(df["State"] == "OH") | (df["State"] == "LA")] 

State  Account length  Area code International plan Voice mail plan  \
1       OH             107        415                 No             Yes   
3       OH              84        408                Yes              No   
8       LA             117        408                 No              No   
32      LA             172        408                 No              No   
91      LA             155        415                 No              No   
...    ...             ...        ...                ...             ...   
3286    OH             106        415                 No             Yes   
3302    LA              75        510                Yes              No   
3311    LA             128        415                 No              No   
3325    OH              78        408                 No              No   
3326    OH              96        415                 No              No   

      Number vmail messages  Total day minutes  Total day calls  \
1                      26.0              161.6              123   
3                       0.0              299.4               71   
8                       0.0              184.5               97   
32                      0.0              212.0              121   
91                      0.0              203.4              100   
...                     ...                ...              ...   
3286                   30.0              220.1              105   
3302                    0.0              153.2               78   
3311                    0.0              147.7               94   
3325                    0.0              193.4               99   
3326                    0.0              106.6              128   

      Total day charge  Total eve minutes  Total eve calls  Total eve charge  \
1                27.47              195.5              103             16.62   
3                50.90               61.9               88              5.26   
8                31.37              351.6               80             29.89   
32               36.04               31.2              115              2.65   
91               34.58              190.9              104             16.23   
...                ...                ...              ...               ...   
3286             37.42              222.2              109             18.89   
3302             26.04              210.8               99             17.92   
3311             25.11              283.3               83             24.08   
3325             32.88              116.9               88              9.94   
3326             18.12              284.8               87             24.21   

      Total night minutes  Total night calls  Total night charge  \
1                   254.4                103               11.45   
3                   196.9                 89                8.86   
8                   215.8                 90                9.71   
32                  293.3                 78               13.20   
91                  196.0                119                8.82   
...                   ...                ...                 ...   
3286                158.4                 96                7.13   
3302                153.5                100                6.91   
3311                188.3                124                8.47   
3325                243.3                109               10.95   
3326                178.9                 92                8.05   

      Total intl minutes  Total intl calls  Total intl charge  \
1                   13.7                 3               3.70   
3                    6.6                 7               1.78   
8                    8.7                 4               2.35   
32                  12.6                10               3.40   
91                   8.9                 4               2.40   
...                  ...               ...                ...   
3286                13.1                 8            

**Какова максимальная длина международных звонков среди лояльных пользователей (`Churn == 0`), не пользующихся услугой международного роуминга (`'International plan' == 'No'`)?**

In [ ]:
df[(df["Churn"] == 0) & (df["International plan"] == "No")]["Total intl minutes"].max()

18.9

#### Встроенные функции

 Могут вычисляться как по строкам, так и по солбцам, как для всего датафрейма, так и для отдельных столбцов

In [ ]:
df.sum(axis=1)

0       1664.76
1       1565.44
2       1486.69
3       1380.60
4       1416.09
         ...   
3328    1657.80
3329    1373.93
3330    1547.34
3331    1607.13
3332    1622.41
Length: 3333, dtype: float64

In [ ]:
#
df.sum(axis=0) 
df.min(axis=1)
df.max()
#среднее
df.mean() 
# медиана
df.median()
# стандартное отклонение
df.std()
# дисперсия
df.var()

**Сколько длился самый долгий дневной звонок?**

In [ ]:
df['Total day minutes'].max()

350.8

Можно написать и применить свою функцию к столбцу

In [8]:
# 1 Способ. Создание функции через def
def minutes_to_hour(x):
    """ Конвертирует минуты в часы
    """
    return x/60

minutes_to_hour(120)

2.0

In [9]:
df2 = df[['State', 'Account length', 'Area code','Total day minutes']].copy()

df2['Hour_duration'] = df2['Total day minutes'].apply(minutes_to_hour)

In [ ]:
df2

State  Account length  Area code  Total day minutes  Hour_duration
0       KS             128        415              265.1       4.418333
1       OH             107        415              161.6       2.693333
2       NJ             137        415              243.4       4.056667
3       OH              84        408              299.4       4.990000
4       OK              75        415              166.7       2.778333
...    ...             ...        ...                ...            ...
3328    AZ             192        415              156.2       2.603333
3329    WV              68        415              231.1       3.851667
3330    RI              28        510              180.8       3.013333
3331    CT             184        510              213.8       3.563333
3332    TN              74        415              234.4       3.906667

[3333 rows x 5 columns]

In [13]:
# 2 Способ. Применение функции через lambda-выражение
df2['Hour_duration_lambda'] = df2['Total day minutes'].apply(lambda x: x/60)
df2

State  Account length  ...  Hour_duration  Hour_duration_lambda
0       KS             128  ...       4.418333              4.418333
1       OH             107  ...       2.693333              2.693333
2       NJ             137  ...       4.056667              4.056667
3       OH              84  ...       4.990000              4.990000
4       OK              75  ...       2.778333              2.778333
...    ...             ...  ...            ...                   ...
3328    AZ             192  ...       2.603333              2.603333
3329    WV              68  ...       3.851667              3.851667
3330    RI              28  ...       3.013333              3.013333
3331    CT             184  ...       3.563333              3.563333
3332    TN              74  ...       3.906667              3.906667

[3333 rows x 6 columns]

#### Удаление данных

In [15]:
# удаление столбцов
df2.drop(columns=['Hour_duration_lambda'], inplace=False)

State  Account length  Area code  Total day minutes  Hour_duration
0       KS             128        415              265.1       4.418333
1       OH             107        415              161.6       2.693333
2       NJ             137        415              243.4       4.056667
3       OH              84        408              299.4       4.990000
4       OK              75        415              166.7       2.778333
...    ...             ...        ...                ...            ...
3328    AZ             192        415              156.2       2.603333
3329    WV              68        415              231.1       3.851667
3330    RI              28        510              180.8       3.013333
3331    CT             184        510              213.8       3.563333
3332    TN              74        415              234.4       3.906667

[3333 rows x 5 columns]

In [16]:
# удаление строк
df.drop(index=[0])

State  Account length  ...  Customer service calls  Churn
1       OH             107  ...                       1  False
2       NJ             137  ...                       0  False
3       OH              84  ...                       2  False
4       OK              75  ...                       3  False
5       AL             118  ...                       0  False
...    ...             ...  ...                     ...    ...
3328    AZ             192  ...                       2  False
3329    WV              68  ...                       3  False
3330    RI              28  ...                       2  False
3331    CT             184  ...                       2  False
3332    TN              74  ...                       0  False

[3332 rows x 20 columns]

In [18]:
# удаление дубликатов
df.drop_duplicates(subset='State', keep='first', inplace=False)

State  Account length  ...  Customer service calls  Churn
0      KS             128  ...                       1  False
1      OH             107  ...                       1  False
2      NJ             137  ...                       0  False
4      OK              75  ...                       3  False
5      AL             118  ...                       0  False
6      MA             121  ...                       3  False
7      MO             147  ...                       0  False
8      LA             117  ...                       1  False
9      WV             141  ...                       0  False
10     IN              65  ...                       4   True
11     RI              74  ...                       0  False
12     IA             168  ...                       1  False
13     MT              95  ...                       3  False
15     NY             161  ...                       4   True
16     ID              85  ...                       1  False
17     VT              93  ...                       3  False
18     VA              76  ...                       1  False
19     TX              73  ...                       1  False
20     FL             147  ...                       0  False
21     CO              77  ...                       5   True
22     AZ             130  ...                       0  False
23     SC             111  ...                       2  False
25     NE             174  ...                       3  False
26     WY              57  ...                       0  False
29     HI              49  ...                       1  False
30     IL             142  ...                       2  False
31     NH              75  ...                       1  False
35     GA              72  ...                       3  False
36     AK              36  ...                       0  False
41     MD             135  ...                       0   True
42     AR              34  ...                       2  False
44     WI              64  ...                       1  False
45     OR              59  ...                       2  False
46     MI              65  ...                       3  False
47     DE             142  ...                       2  False
53     UT              96  ...                       2  False
61     CA             116  ...                       2  False
62     MN              74  ...                       2  False
63     SD             149  ...                       3  False
64     NC              38  ...                       2  False
65     WA              40  ...                       2  False
72     NM             147  ...                       3  False
73     NV              90  ...                       1  False
76     DC              82  ...                       0   True
88     KY              74  ...                       2  False
93     ME              78  ...                       3  False
98     MS             120  ...                       1   True
196    TN              78  ...                       0  False
212    PA              82  ...                       1  False
290    CT             199  ...                       0  False
377    ND              79  ...                       1  False

[51 rows x 20 columns]

### Группировка данных

В общем случае группировка данных в Pandas выглядит следующим образом:

```
df.groupby(by=grouping_columns)[columns_to_show].function()
```

1. К датафрейму применяется метод **`groupby`**, который разделяет данные по `grouping_columns` – признаку или набору признаков.
3. Индексируем по нужным нам столбцам (`columns_to_show`). 
2. К полученным группам применяется агрегирующая функция или несколько функций.

In [ ]:
# Группировка по одному полю
df.groupby('Churn').count()

State  Account length  ...  Total intl charge  Customer service calls
Churn                         ...                                           
False   2850            2850  ...               2850                    2850
True     483             483  ...                483                     483

[2 rows x 19 columns]

In [ ]:
df.groupby('Churn')[['State']].count()

State
Churn       
False   2850
True     483

In [6]:
# Переименуем столбец
df.groupby('Churn')[['State']].count().rename(columns={'State':'Number of abonents'})

Number of abonents
Churn                    
False                2850
True                  483

In [ ]:
df.groupby('State')[['Total day charge']].sum()

Total day charge
State                  
AK              1576.94
AL              2529.72
AR              1646.70
AZ              1867.08
CA              1061.01
CO              2005.20
CT              2203.30
DC              1573.31
DE              1810.49
FL              1922.85
GA              1698.58
HI              1585.39
IA              1328.60
ID              2216.70
IL              1711.67
IN              2372.11
KS              2279.56
KY              1742.81
LA              1546.56
MA              1990.14
MD              2347.07
ME              1952.74
MI              2241.17
MN              2618.36
MO              1826.20
MS              1966.15
MT              2011.55
NC              2140.32
ND              1974.61
NE              1840.33
NH              1688.19
NJ              2268.39
NM              1806.86
NV              1979.53
NY              2470.94
OH              2430.29
OK              1865.67
OR              2337.10
PA              1441.10
RI              1850.69
SC              1697.75
SD              1934.86
TN              1583.73
TX              2221.81
UT              2246.92
VA              2320.14
VT              2259.08
WA              2005.52
WI              2375.28
WV              3134.67
WY              2358.43

In [ ]:
# Группировка по нескольким полям
df.groupby(['Churn', 'International plan'])[['Total day minutes']].count()

Total day minutes
Churn International plan                   
False No                               2664
      Yes                               186
True  No                                346
      Yes                               137

In [ ]:
# агрегирующие функции
'''
count() - количество значений в группе
first(), last() - первое и последнее значение в группе
mean(), median() - среднее арифметическое и медиана
min(), max() - минимальное и максимальное значения
std(), var() - стандартное отклонение и дисперсия
mad() - среднее абсолютное отклонение 
sum() - сумма
prod() - произведение
describe() - расчет всех описательных статистик
'''

'\ncount() - количество значений в группе\nfirst(), last() - первое и последнее значение в группе\nmean(), median() - среднее арифметическое и медиана\nmin(), max() - минимальное и максимальное значения\nstd(), var() - стандартное отклонение и дисперсия\nmad() - среднее абсолютное отклонение \nsum() - сумма\nprod() - произведение\ndescribe() - расчет всех описательных статистик\n'

In [ ]:
columns_to_show = ["Total day minutes", "Total eve minutes", "Total night minutes"]

df.groupby(["Churn"])[columns_to_show].describe()

Total day minutes                         ... Total night minutes               
                  count        mean        std  ...                 50%     75%    max
Churn                                           ...                                   
False            2850.0  175.175754  50.181655  ...              200.25  234.90  395.0
True              483.0  206.914079  68.997792  ...              204.80  239.85  354.9

[2 rows x 24 columns]

Можно применять несколько агрегирующиъ функций в одной группировке

In [ ]:
columns_to_show = ["Total day minutes", "Total eve minutes", "Total night minutes"]

df.groupby(["Churn"])[columns_to_show].agg(['min', 'max', 'count'])

Total day minutes               ... Total night minutes             
                    min    max count  ...                 min    max count
Churn                                 ...                                 
False               0.0  315.6  2850  ...                23.2  395.0  2850
True                0.0  350.8   483  ...                47.4  354.9   483

[2 rows x 9 columns]

Можно так же к каждому столбцу применить свою агрегирующую функцию

In [ ]:
df.groupby(["Churn"]).agg({"Total day minutes": 'sum', 'Total eve minutes':['min','max','mean']})

Total day minutes Total eve minutes                   
                    sum               min    max        mean
Churn                                                       
False          499250.9               0.0  361.8  199.043298
True            99939.5              70.9  363.7  212.410145

In [ ]:
# Не забывайте присваивать новой переменной результат группировки!
tmp = df.groupby(["Churn"]).agg({"Total day minutes": 'sum', 'Total eve minutes':['min','max','mean']})
tmp

Total day minutes Total eve minutes                   
                    sum               min    max        mean
Churn                                                       
False          499250.9               0.0  361.8  199.043298
True            99939.5              70.9  363.7  212.410145

Часто удобно использовать метод **`value_counts`**, который выполняет группировку и подсчет количества, а также вывод данные в отсортированном по убыванию виде

In [ ]:
df['State'].value_counts()

Такой же результат, как в ячейке выше, получим при использовании группировки. Но кода здесь больше

In [ ]:
df.groupby('State')[['Churn']].count().rename(columns={'Churn':'Count'}).sort_values('Count',ascending=False)

Count
State       
WV       106
MN        84
NY        83
AL        80
OR        78
WI        78
OH        78
VA        77
WY        77
CT        74
VT        73
ID        73
MI        73
UT        72
TX        72
IN        71
KS        70
MD        70
NJ        68
NC        68
MT        68
CO        66
NV        66
WA        66
RI        65
MS        65
MA        65
AZ        64
MO        63
FL        63
ND        62
ME        62
NM        62
OK        61
NE        61
DE        61
SC        60
SD        60
KY        59
IL        58
NH        56
AR        55
DC        54
GA        54
HI        53
TN        53
AK        52
LA        51
PA        45
IA        44
CA        34

**Посмотрим распределение значений по полую `Churn`**

In [ ]:
df["Churn"].value_counts()

False    2850
True      483
Name: Churn, dtype: int64

2850 пользователей из 3333 — лояльные, значение переменной `Churn` у них — `0`.

Посмотрим на распределение пользователей по переменной `Area code`. Укажем значение параметра `normalize=True`, чтобы посмотреть не абсолютные частоты, а относительные.

In [ ]:
df["Area code"].value_counts(normalize=True)

415    0.496550
510    0.252025
408    0.251425
Name: Area code, dtype: float64

### Пропуски в данных

**Поиск пропусков**

Иногда в наборах данных возможно пропуски значений. Проверим, есть ли такие в нашем наборе можно несколькими способами.

Можно сравнить количесво значений по столбцам через метод `describe()` или же с использованием специального метода `isna()`

In [ ]:
df.describe()

Account length    Area code  ...  Total intl charge  Customer service calls
count     3333.000000  3333.000000  ...        3333.000000             3333.000000
mean       101.064806   437.182418  ...           2.764581                1.562856
std         39.822106    42.371290  ...           0.753773                1.315491
min          1.000000   408.000000  ...           0.000000                0.000000
25%         74.000000   408.000000  ...           2.300000                1.000000
50%        101.000000   415.000000  ...           2.780000                1.000000
75%        127.000000   510.000000  ...           3.270000                2.000000
max        243.000000   510.000000  ...           5.400000                9.000000

[8 rows x 16 columns]

In [ ]:
# Проверяем, в каких столбцах есть пропуски
df.isna().any()

State                     False
Account length            False
Area code                 False
International plan        False
Voice mail plan           False
Number vmail messages      True
Total day minutes         False
Total day calls           False
Total day charge          False
Total eve minutes         False
Total eve calls           False
Total eve charge           True
Total night minutes       False
Total night calls         False
Total night charge        False
Total intl minutes        False
Total intl calls          False
Total intl charge         False
Customer service calls    False
Churn                     False
dtype: bool

In [ ]:
# маска значений по одному столбцу
df['Number vmail messages'].isna() #isnull

0       False
1       False
2       False
3       False
4       False
        ...  
3328    False
3329    False
3330    False
3331    False
3332    False
Name: Number vmail messages, Length: 3333, dtype: bool

In [ ]:
# фильтрация датафрейма по пустым значениям в столбце
df[df['Number vmail messages'].isna()]

State  Account length  Area code International plan Voice mail plan  \
24    VA             132        510                 No              No   
53    UT              96        415                 No              No   

    Number vmail messages  Total day minutes  Total day calls  \
24                    NaN               81.1               86   
53                    NaN              160.2              117   

    Total day charge  Total eve minutes  Total eve calls  Total eve charge  \
24             13.79              245.2               72             20.84   
53             27.23              267.5               67             22.74   

    Total night minutes  Total night calls  Total night charge  \
24                237.0                115               10.67   
53                228.5                 68               10.28   

    Total intl minutes  Total intl calls  Total intl charge  \
24                10.3                 2               2.78   
53                 9.3                 5               2.51   

    Customer service calls  Churn  
24                       0      0  
53                       2      0

In [ ]:
df.loc[:,:]

In [ ]:
# отфильтровать весь датафрейм не получится - получите все значения NaN
df[df.isna()]

State  Account length  Area code International plan Voice mail plan  \
0      NaN             NaN        NaN                NaN             NaN   
1      NaN             NaN        NaN                NaN             NaN   
2      NaN             NaN        NaN                NaN             NaN   
3      NaN             NaN        NaN                NaN             NaN   
4      NaN             NaN        NaN                NaN             NaN   
...    ...             ...        ...                ...             ...   
3328   NaN             NaN        NaN                NaN             NaN   
3329   NaN             NaN        NaN                NaN             NaN   
3330   NaN             NaN        NaN                NaN             NaN   
3331   NaN             NaN        NaN                NaN             NaN   
3332   NaN             NaN        NaN                NaN             NaN   

      Number vmail messages  Total day minutes  Total day calls  \
0                       NaN                NaN              NaN   
1                       NaN                NaN              NaN   
2                       NaN                NaN              NaN   
3                       NaN                NaN              NaN   
4                       NaN                NaN              NaN   
...                     ...                ...              ...   
3328                    NaN                NaN              NaN   
3329                    NaN                NaN              NaN   
3330                    NaN                NaN              NaN   
3331                    NaN                NaN              NaN   
3332                    NaN                NaN              NaN   

      Total day charge  Total eve minutes  Total eve calls  Total eve charge  \
0                  NaN                NaN              NaN               NaN   
1                  NaN                NaN              NaN               NaN   
2                  NaN                NaN              NaN               NaN   
3                  NaN                NaN              NaN               NaN   
4                  NaN                NaN              NaN               NaN   
...                ...                ...              ...               ...   
3328               NaN                NaN              NaN               NaN   
3329               NaN                NaN              NaN               NaN   
3330               NaN                NaN              NaN               NaN   
3331               NaN                NaN              NaN               NaN   
3332               NaN                NaN              NaN               NaN   

      Total night minutes  Total night calls  Total night charge  \
0                     NaN                NaN                 NaN   
1                     NaN                NaN                 NaN   
2                     NaN                NaN                 NaN   
3                     NaN                NaN                 NaN   
4                     NaN                NaN                 NaN   
...                   ...                ...                 ...   
3328                  NaN                NaN                 NaN   
3329                  NaN                NaN                 NaN   
3330                  NaN                NaN                 NaN   
3331                  NaN                NaN                 NaN   
3332                  NaN                NaN                 NaN   

      Total intl minutes  Total intl calls  Total intl charge  \
0                    NaN               NaN                NaN   
1                    NaN               NaN                NaN   
2                    NaN               NaN                NaN   
3                    NaN               NaN                NaN   
4                    NaN               NaN                NaN   
...                  ...               ...                ...   
3328                 NaN               NaN            

Пропуски можно **удалить** из данных либо чем-то **заполнить**, например, нулем, или каким-то другим значением. Чем заполнять - зависит от конкретного набора даннх и задачи

In [ ]:
# удаление строк, в которых есть хть один пропуск
df = df.dropna() #inplace=True

State  Account length  ...  Customer service calls  Churn
0       KS             128  ...                       1  False
1       OH             107  ...                       1  False
2       NJ             137  ...                       0  False
3       OH              84  ...                       2  False
4       OK              75  ...                       3  False
...    ...             ...  ...                     ...    ...
3328    AZ             192  ...                       2  False
3329    WV              68  ...                       3  False
3330    RI              28  ...                       2  False
3331    CT             184  ...                       2  False
3332    TN              74  ...                       0  False

[3329 rows x 20 columns]

In [ ]:
tmp = pd.DataFrame({'a':[1,2,3], 'b':[3, np.nan, 5]})
tmp.fillna(0, inplace=True)
tmp

a    b
0  1  3.0
1  2  0.0
2  3  5.0

In [ ]:
# заполнение пропусков нулем
df.fillna(0) #inplace=True
df

State  Account length  ...  Customer service calls  Churn
0       KS             128  ...                       1  False
1       OH             107  ...                       1  False
2       NJ             137  ...                       0  False
3       OH              84  ...                       2  False
4       OK              75  ...                       3  False
...    ...             ...  ...                     ...    ...
3328    AZ             192  ...                       2  False
3329    WV              68  ...                       3  False
3330    RI              28  ...                       2  False
3331    CT             184  ...                       2  False
3332    TN              74  ...                       0  False

[3333 rows x 20 columns]